### ISSUE TITLE
clojure.test predicate linters [issue needs splitting up]

### ISSUE THREAD

**Issue by vemv** [clj-kondo/clj-kondo#893]
## Is your feature request related to a problem? Please describe.

clojure.test tests, when failing, should fail in a graceful, informative way that effectively conveys the difference between 'expected' and 'actual'.

clojure.test cannot create great failure reports for every kind of test. However, if those same tests are rewritten in specific ways, they can have nice reporting again.

## Describe the solution you'd like

clj-kondo could lint when it is detected that a form is not ideal _and_ there's a known fix for it.  

These are some faulty forms that come to mind.

One can verify for each of them that a failing test would fail in an opaque manner.

I'll use diff notation from 'bad' to 'good' forms.

### expected/actual order

```diff
- (is (= (actual expected)))
+ (is (= (expected actual)))
```

Of course, a computer cannot guess what one's expectation is. However the following heuristics apply: 

* if one hand of the equality is prefixed with [`sut/`](https://en.wikipedia.org/wiki/System_under_test), that's the expected part.
  * so, `(= foo (sut/bar))` is wrong.
* if one hand of the equality is a literal value, that's the actual part
  * so, `(= 2 foo)` is wrong.

There may be other heuristics. Of course, there will be cases where no heuristic can be found.

⚠️ Right now I cannot come up with an example of a bad report caused by this, but it's something that certainly tends to bite people.
One can find evidence of this specific order convention in the existence of https://github.com/magnars/kaocha-noyoda (e.g. if there wasn't a specific conventional order, that plugin wouldn't exist). 

### `->`

```diff
- (is (-> a :b :c foo?))
+ (let [v (-> a :b :c)]
+   (is (foo? v)))
```

### `not nil`

```diff
- (is (not (nil? v)))
+ (is (some? v))
```

### `every?` / `not-any?`

```diff
- (is (every? odd? xs))
+ (doseq [x xs] (is (odd? x)))
```

Same for `not-any?`.

### anonymous functions

```diff
- (is (some #(foo? % :arg) xs)) 
+ (is (some (fn pred-name [x] (foo? x :arg)) xs))
```

Emphasis on `pred-name`. Else anonymous functions will not be printed readably on reports.

### `and`

```diff
- (is (and x y))
+ (is x)
+ (is y)
```

## Describe alternatives you've considered

None that I could come up w/

## Additional context

Here are the same lintable cases in runnable form.

```clj
(deftest plenty-of-badness

  (let [a {}
        foo? (fn [x]
               (= x 42))]
    (is (-> a :b :c foo?)
        "wrong")

    (let [v (-> a :b :c)]
      (is (foo? v)
          "better (it tells me that `(foo? v)` was false, instead of telling me that the whole `->` was wrong (which part of it?))")))

  (let [v nil]
    (is (not (nil? v))
        "wrong or ugly (humans tend do dislike double negatives; see the reported `not not`)")

    (is (some? v)
        "better"))

  (let [xs [1 2 3]]
    (is (every? odd? xs)
        "wrong: doesn't tell you the culprit collection value")

    (doseq [x xs]
      (is (odd? x)
          "better")))

  (let [xs [1 2 3]]
    (is (not-any? odd? xs)
        "wrong: doesn't tell you the culprit collection value")

    (doseq [x xs]
      (is (not (odd? x))
          "better")))

  (let [foo? odd?
        xs [2 4 6]]
    (is (some #(foo? %) xs)
        "wrong")
    
    (is (some (fn pred-name [x]
                (foo? x))
              xs)
        "better (although marginally): now there's `pred-name` in the report. Useful if a given test contains _multiple_ fns (i.e. **which** of them failed?)"))

  (let [x 1
        y nil]
    (is (and x y)
        "wrong: doesn't tell you which of `[x, y]` evaluated to false")))
```

**Comment by borkdude** [comment]
There are lots of opinionated things in this issue that makes it hard to implement this issue in one go. Also there might be things that are not specific to testing. E.g.:

```
- (not (nil? v))
+ (some? v)
```

My suggestion would be to create smaller issues out of these so we can discuss each item separately.

**Comment by borkdude** [comment]
> so, (= 2 foo) is wrong

2 is the expected value. foo is the actual value. so, that should be good right?

**Comment by borkdude** [comment]
I'll re-open it and put it in the triage column. If people want to help out, they can create separate issues out of this (which should then again be triaged individually).

**Comment by vemv** [comment]
> 2 is the expected value. foo is the actual value. so, that should be good right?

Sorry, yes, I wrote that one down too quickly. Edited.

**Comment by vemv** [comment]
I can reopen this issue properly split at some point later.

**Comment by borkdude** [comment]
New (more focused) issue welcome
